cookbooks/anisotropic_viscosity/av_material.cc

```diff
@@ -0,0 +1,885 @@
+      AssertThrow((dim==2),
+                  ExcMessage("For now the anisotropic viscosity assemblers work only in 2D"));
+
+    }
+
+
+
+    template <int dim>
+    void
+    AV<dim>::evaluate (const MaterialModel::MaterialModelInputs<dim> &in,
+                       MaterialModel::MaterialModelOutputs<dim> &out) const
+    {
+      MaterialModel::AnisotropicViscosity<dim> *anisotropic_viscosity =
+        out.template get_additional_output<MaterialModel::AnisotropicViscosity<dim> >();
+
+      Simple<dim>::evaluate(in,out);
+
+      AssertThrow((this->introspection().compositional_name_exists("gamma")),
+                  ExcMessage("AV material model only works if there is a compositional field called gamma."));
+      AssertThrow(this->introspection().compositional_name_exists("ni"),
+                  ExcMessage("AV material model only works if there is a compositional field called ni."));
+      AssertThrow(this->introspection().compositional_name_exists("nj"),
+                  ExcMessage("AV material model only works if there is a compositional field called nj."));
+      if (dim == 3)
+        AssertThrow(this->introspection().compositional_name_exists("nk"),
+                    ExcMessage("AV material model only works if there is a compositional field called nk."));
+
+      const unsigned int c_idx_gamma = this->introspection().compositional_index_for_name("gamma");
+
+      std::vector<unsigned int> c_idx_n;
+      c_idx_n.push_back (this->introspection().compositional_index_for_name("ni"));
+      c_idx_n.push_back (this->introspection().compositional_index_for_name("nj"));
+      if (dim == 3)
+        c_idx_n.push_back (this->introspection().compositional_index_for_name("nk"));
+
+      // Get the grad_u tensor, at the center of this cell, if possible.
+
+      std::vector<Tensor<2,dim> > velocity_gradients (in.position.size());
+      if (in.current_cell.state() == IteratorState::valid)
+        {
+          std::vector<Point<dim> > quadrature_positions(in.position.size());
+          for (unsigned int i=0; i < in.position.size(); ++i)
+            quadrature_positions[i] = this->get_mapping().transform_real_to_unit_cell(in.current_cell, in.position[i]);
+
+          // FEValues requires a quadrature and we provide the default quadrature
+          // as we only need to evaluate the gradients of the solution.
+          FEValues<dim> fe_values (this->get_mapping(),
+                                   this->get_fe(),
+                                   Quadrature<dim>(quadrature_positions),
+                                   update_gradients);
+          fe_values.reinit (in.current_cell);
+
+
+          fe_values[this->introspection().extractors.velocities]
+          .get_function_gradients(this->get_solution(), velocity_gradients);
+
+        }
+
+      for (unsigned int q=0; q<in.position.size(); ++q)
+        {
+          out.densities[q] = (in.composition[q][c_idx_gamma] > 0.8 ? 1 : 0);
```

Comment:
Why do you need to set densities etc here and in the following lines? Don't you inherit these values from the `Simple` model that you call above via `Simple<dim>::evaluate(in, out)`?